From david-b@pacbell.net  Fri Jun 27 15:12:21 2008
From: Robert Jarzmik <rjarzmik@free.fr>
Date: Tue, 10 Jun 2008 15:01:15 -0700
Subject: USB: Make pxa27x_udc less invasive on UP2OCR pxa register.
To: Robert Jarzmik <rjarzmik@free.fr>
Cc: linux-usb@vger.kernel.org
Message-ID: <200806101501.15282.david-b@pacbell.net>
Content-Disposition: inline

From: Robert Jarzmik <rjarzmik@free.fr>

UP2OCR registers sets up usb client function, as well as
other board specific setups (resistors pullups and
pulldowns). Let the driver setup the "non-OTG client" part,
and let all other bits as they were set by board code.

Signed-off-by: Robert Jarzmik <rjarzmik@free.fr>
Acked-by: David Brownell <dbrownell@users.sourceforge.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/pxa27x_udc.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

--- a/drivers/usb/gadget/pxa27x_udc.c
+++ b/drivers/usb/gadget/pxa27x_udc.c
@@ -1573,9 +1573,16 @@ static __init void udc_init_data(struct 
  */
 static void udc_enable(struct pxa_udc *udc)
 {
+	u32 up2ocr;
+
 	udc_writel(udc, UDCICR0, 0);
 	udc_writel(udc, UDCICR1, 0);
-	udc_writel(udc, UP2OCR, UP2OCR_HXOE);
+	/*
+	 * UP2OCR : set HXOE=1, HXS=0, SEOS=0 => non-OTG client
+	 */
+	up2ocr = udc_readl(udc, UP2OCR);
+	up2ocr = (up2ocr & ~(UP2OCR_HXS - 1)) | UP2OCR_HXOE;
+	udc_writel(udc, UP2OCR, up2ocr);
 	udc_clear_mask_UDCCR(udc, UDCCR_UDE);
 
 	clk_enable(udc->clk);
